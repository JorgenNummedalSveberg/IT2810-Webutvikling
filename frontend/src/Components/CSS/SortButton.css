@import url('https://fonts.googleapis.com/css?family=Roboto&display=swap');

.SortButton{
    display: flex;
    justify-content: center;
    border: 0;
    padding: 0;
    max-width: 120px;
    width: 25%;
    height: auto;
    min-width: none;
    overflow: hidden;
    cursor: pointer;
}

.SortButton svg{
    height: 100%;
    width: 9%;
    min-width: 5px;
    -webkit-filter: drop-shadow( 3px 3px 5px rgba(0, 0, 0, 0.5));
    filter: drop-shadow( 3px 3px 5px rgba(0, 0, 0, 0.5));
    margin-left: 10px;
    transition: all 0.5s ease, ease-out 0.5s;
}

.SortButton h3{
    font-family: "Roboto";
    user-select: none;
    margin: auto;
    margin-left: 10px;
    margin-right: 20px;
    font-size: 150%;
    text-shadow: 3px 3px 5px rgba(0, 0, 0, 0.4);
    transition: all 0.5s ease, ease-out 0.5s;
}

.SortButton:hover h3, svg{
    text-shadow: 3px 3px 10px rgba(0, 0, 0, 0.6);
}

.SortButton:hover svg{
    filter: drop-shadow( 3px 3px 1px rgba(0, 0, 0, 0.4));
}
