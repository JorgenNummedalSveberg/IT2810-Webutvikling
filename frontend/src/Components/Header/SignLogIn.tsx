import React, {useState} from "react";
import {useDispatch} from "react-redux";
import {User} from "../../types/User";
import {login, logout} from "../../actions";
import {Button, Dialog, DialogActions, DialogContent, DialogTitle, Input} from "@material-ui/core";
import {makeStyles} from "@material-ui/styles";

function SignLogIn(props: { isLogged: boolean }) {

    // Nødvendig for redux
    const dispatch = useDispatch();

    // Holder styr på inputs
    const [userName, setUsername] = useState("");
    const [password, setPassword] = useState("");
    const [open, setOpen] = useState(false);

    // Holder styr på error
    const [error, setError] = useState<false | { message: string, log: boolean }>(false);

    // Lager request for fetch
    const req = (reqUser: User) => {
        return ({
            method: 'POST',
            body: JSON.stringify(reqUser),
            headers: {
                'Content-Type': 'application/json'
            }
        })
    }

    // Legger til en bruker på serveren
    function addUser(reqUser: User) {
        setOpen(false);
        fetch('http://localhost:5000/api/user/add', req(reqUser))
            .then(response => {
                if (response.ok) {
                    dispatch(login(reqUser));
                } else {
                    setError({message: "Username is taken", log: false});
                }
            })
    }

    // Logger inn hvis brukeren finnes
    function onLogin(user: User) {
        setOpen(false);
        fetch('http://localhost:5000/api/user?userName=' + user.userName + '&password=' + user.password)
            .then(response => {
                if (response.ok) {
                    response.json()
                        .then(movies => {
                            user.movies = movies;
                            dispatch(login(user))
                        })
                } else {
                    setError({message: "Username or password is wrong", log: true});
                }
            })
    }

    // @ts-ignore
    function handleNameChange(value: string) {
        setError(false);
        setUsername(value);
    }

    // @ts-ignore
    function handlePasswordChange(value: string) {
        setError(false);
        setPassword(value)
    }

    const classes = makeStyles({
        loginButton: {
            backgroundColor: '#E85A4F',
            borderRadius: '5px',
            margin: '10px',
            '& span': {
                fontSize: '1.8em',
                color: 'white',
            }
        },
        none: {
            display: 'none'
        },
        initial: {
            display: 'initial'
        }
    })
    return (
        <div>
            <div className={`${props.isLogged ? classes().initial : classes().none}`}>
                <Button className={classes().loginButton} onClick={() => setOpen(true)}>Log in/Sign up</Button>
                <Dialog open={open} onClose={() => setOpen(false)} title='Log in/Sign up'>
                    <DialogTitle>Log in/Sign up</DialogTitle>
                    <DialogContent>
                        <Input error={!!error} id={"UsernameID"} autoFocus title={"Username"}
                               onChange={(e) => handleNameChange(e.target.value)}
                               name={"userName"} placeholder='Username'/>
                    </DialogContent>
                    <DialogContent>
                        <Input type={"password"} error={!!error && error.log} id={"PasswordID"} title={"Password"}
                               onChange={(e) => handlePasswordChange(e.target.value)}
                               name={"password"} placeholder='Password'/>
                    </DialogContent>
                    <DialogActions>
                        <Button id={"loginButtonID"} onClick={() => onLogin({
                            userName: userName,
                            password: password,
                            movies: []
                        })} type='submit'>Log in</Button>
                        <Button id={"submitButtonID"}
                                onClick={() => addUser({userName: userName, password: password, movies: []})}
                                type='submit'>Sign
                            up</Button>
                    </DialogActions>
                </Dialog>
            </div>
            <div className={`${props.isLogged ? classes().none : classes().initial}`}>
                <Button onClick={() => dispatch(logout())} className={classes().loginButton}>Log out</Button>
            </div>
        </div>
    )
}

export default SignLogIn;