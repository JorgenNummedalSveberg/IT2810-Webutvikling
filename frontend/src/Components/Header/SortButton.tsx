import React from 'react';
import {useDispatch, useSelector} from "react-redux";
import {setDesc, setSort} from "../../actions";
import {State} from "../../types/State";
import {makeStyles} from "@material-ui/styles";
import KeyboardArrowDownIcon from '@material-ui/icons/KeyboardArrowDown';
import KeyboardArrowUpIcon from '@material-ui/icons/KeyboardArrowUp';
import KeyboardArrowLeftIcon from '@material-ui/icons/KeyboardArrowLeft';
import {Paper, useTheme} from "@material-ui/core";

// Knapp som oppdaterer hva kategori vi sorterer etter og hvilken retning vi sorterer i
function SortButton(props: { mobile: boolean, sort: string, refresh: () => void, nummer: string }) {

    // Nødvendig for redux
    const dispatch = useDispatch();

    // Definerer om knappen er trykket basert på om filteret i redux state er likt som navnet til denne knappen
    const active = useSelector((state: State) => state.filter.sort) === props.sort;
    // Bestemmer rentning basert på filter i state
    const desc = useSelector((state: State) => state.filter.desc);

    // Håndterer trykk på knapp, endrer filter i state basert på tittel og retning
    function toggleSort() {
        if (active) {
            // Hvis aktiv, bare bytt retning
            dispatch(setDesc(!desc));
        } else {
            // Hvis ikke aktiv, sett sort i state og set descending til true
            dispatch(setSort(props.sort));
            dispatch(setDesc(true));
        }
        props.refresh();
    }
    const theme = useTheme();
    const classes = makeStyles({
        div: {
            display: 'flex',
            flexDirection: props.mobile ? 'column' : 'row',
            alignItems: 'flex-start',
            textAlign: 'center',
            backgroundColor: theme.palette.primary.main,
            '&:hover': {
                backgroundColor: theme.palette.primary.dark
            },
            borderRadius: '5px',
            cursor: 'pointer',
            margin: '10px',
            height: 'calc(100%-20px)'
        },
        sortTitle: {
            color: active ? theme.palette.getContrastText(theme.palette.primary.light) : theme.palette.text.disabled,
            width: '100%',
            margin: 0,
            paddingTop: props.mobile ? '40px' : '7px'
        },
        arrow: {
            color: active ? theme.palette.getContrastText(theme.palette.primary.light) : theme.palette.text.disabled
        }
    })

    return (
        <Paper elevation={5} className={classes().div} data-testid={"sortbutton" + props.nummer} id={"sortbutton" + props.nummer}
               onClick={toggleSort}>
            {props.mobile ? <h3 className={classes().sortTitle}>{props.sort}</h3> : null}
            {active ? (desc ?
                <KeyboardArrowDownIcon className={classes().arrow}/> :
                <KeyboardArrowUpIcon className={classes().arrow}/>) :
                <KeyboardArrowLeftIcon className={classes().arrow}/>}
            {!props.mobile ? <h3 className={classes().sortTitle}>{props.sort}</h3> : null}
        </Paper>
    );
}

export default SortButton;
