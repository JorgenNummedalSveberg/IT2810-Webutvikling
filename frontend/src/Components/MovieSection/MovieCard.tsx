import {Movie} from "../../types/Movie";
import {useDispatch} from "react-redux";
import {setPopup, showPopup} from "../../actions";
import {
    ButtonBase,
    Card,
    CardContent,
    CardHeader,
    CardMedia,
    Divider,
    Grid,
    Paper,
    useMediaQuery
} from "@material-ui/core";
import {Skeleton} from "@material-ui/lab";
import HourglassFullIcon from '@material-ui/icons/HourglassFull';
import ImdbIcon from "../Shared/ImdbIcon";
import React from "react";
import {makeStyles} from "@material-ui/styles";

// Komponent for å vise frem en film i et kort
function MovieCard(props: { movie: Movie }) {

    // Nødvendig for redux
    const dispatch = useDispatch();

    // Åpner opp Popup.
    function handleClick() {
        dispatch(setPopup(props.movie));
        dispatch(showPopup(true));
    }

    const classes = makeStyles({
        card: {height: "100%", width: '100%', backgroundColor: '#464646'},
        gridItem: {flexGrow:1, flexBasis: 1, maxWidth: '600px', width: '600px'},
        paperButton: {height: "100%", width: '100%'},
        paper: {backgroundColor: '#E85A4F', height: "100%", width: '100%', display: 'flex', flexDirection: 'row'},
        poster: {maxWidth: '500px', minWidth: '250px'},
        details: {padding: '10px', display: 'flex', flexDirection: 'column'},
        title: {flexGrow: 1},
        description: {flexGrow: 4, color: 'white', textAlign: 'left'},
        bottomInfo: {
            margin: '10px',
            display: 'flex',
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'space-between',
            flexGrow: 1,
            '& *': {margin: '5px'}
        },
        duration: {color: '#e5dfca', display: 'flex', alignItems: 'center'},
        noMargin: {
            margin: 0
        }
    })

    return (
        <Grid className={classes().gridItem} item>
            <ButtonBase className={classes().paperButton} onClick={handleClick}>
                <Paper className={classes().paper} elevation={5}>
                    <img className={classes().poster} alt='Could not display movie poster' width='100%' src={props.movie.posterurl}/>
                    <div className={classes().details}>
                        <div className={classes().title}>
                            <h1>{props.movie.title}</h1>
                            <Divider/>
                        </div>
                        <div className={classes().description}>
                            <h3>Year: {props.movie.year}</h3>
                            <h3>Genres: {props.movie.genres.join(', ')}</h3>
                        </div>
                        <div className={classes().bottomInfo}>
                            <div className={classes().duration}>
                                <HourglassFullIcon className={classes().noMargin} fontSize={"large"}/>
                                <p className={classes().noMargin}>{parseTime(props.movie.duration)}</p>
                            </div>
                            <Divider orientation='vertical'/>
                            <ImdbIcon rating={props.movie.imdbRating} height={35}/>
                        </div>
                    </div>
                </Paper>
            </ButtonBase>
        </Grid>

    )
}

// Blanke kort for når nettsiden laster inn filmene
export function DimCard() {
    return (
        <Grid item>
            <Card style={{height: "100%", width: '100%', backgroundColor: '#464646'}}>
                <Skeleton style={{backgroundColor: '#222222'}} variant="rect" width={"100%"} height={300} />
                <Skeleton animation="wave" width={340} height={40} style={{ margin: 10 }} />
                <Skeleton animation="wave" width={340} height={40} style={{ margin: 10 }} />
            </Card>
        </Grid>
    )
}

// Tar inn tid i formatet på databasen og gjør det om til presentabel string, eller rein minuttverdi
function parseTime(time: number): string {
    let minutes = time;
    let hours = 0;
    while (minutes - 60 > 0) {
        minutes -= 60;
        hours++;
    }
    return hours + 'h' + minutes + 'm';
}

export default MovieCard