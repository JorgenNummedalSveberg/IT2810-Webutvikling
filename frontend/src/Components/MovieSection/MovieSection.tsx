import {Button, Drawer, Grid} from "@material-ui/core";
import {showPopup} from "../../actions";
import ArrowBackIcon from "@material-ui/icons/ArrowBack";
import Popup from "./Popup";
import React from "react";
import ErrorPage from "./ErrorPage";
import PagePicker from "./PagePicker";
import PopupContainer from "./PopupContainer";

export default function MovieSection(props: {
    page: number,
    pages: number
    classes: any,
    dispatch: (f: { payload: any; type: string }) => void,
    popupShow: boolean,
    refresh: (number: number) => void,
    movieCards: JSX.Element[]}) {
    return (
        <div className={props.classes().root}>
            <ErrorPage classes={
                {errorPage: props.classes().errorPage,
                    details: props.classes().details,
                    gridItem: props.classes().gridItem,
                    title: props.classes().title,
                    description: props.classes().description,
                    bottomInfo: props.classes().bottomInfo}}/>
            <div className={props.classes().moviePage}>
                <Drawer anchor={'right'} open={props.popupShow} onClose={() => props.dispatch(showPopup(false))}>
                    <Button startIcon={<ArrowBackIcon/>} onClick={() => props.dispatch(showPopup(false))}>Close</Button>
                    <div className={props.classes().popup}>
                        <PopupContainer refresh={props.refresh}/>
                    </div>
                </Drawer>
                <div className={props.classes().main}>
                    <PagePicker refresh={props.refresh} dispatch={props.dispatch} page={props.page} pages={props.pages}/>
                    <Grid
                        className={props.classes().movieGrid}
                        container
                        justify="center"
                        alignItems="stretch"
                        spacing={4}
                    >
                        {props.movieCards}
                    </Grid>
                    <PagePicker refresh={props.refresh} dispatch={props.dispatch} page={props.page} pages={props.pages}/>
                </div>
            </div>
        </div>
    )
}