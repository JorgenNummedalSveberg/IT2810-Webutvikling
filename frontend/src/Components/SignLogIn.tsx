import React, {useState} from "react";
import {Button, Input, Modal} from "semantic-ui-react";
import {useDispatch} from "react-redux";
import {User} from "../types/user";
import {login} from "../actions";

function SignLogIn () {

    // Nødvendig for redux
    const dispatch = useDispatch();

    // Holder styr på inputs
    const [userName, setUsername] = useState("");
    const [password, setPassword] = useState("");

    // Lager request for fetch
    const req = (reqUser: User) => {
        return ({
            method: 'POST',
            body: JSON.stringify(reqUser),
            headers: {
                'Content-Type': 'application/json'
            }
        })
    }

    // Legger til en bruker på serveren
    function addUser(reqUser: User) {
        fetch('http://localhost:5000/api/user/add', req(reqUser))
            .then(response => response.json())
            .then((user: User) => {
                if (!!user) {
                    dispatch(login(reqUser));
                }
            })
            .catch(() => {
                console.log("User already exists")
            });
    }

    // Logger inn hvis brukeren finnes
    function onLogin(user: User) {
        let returnUser = user;
        try {
            fetch('http://localhost:5000/api/user?userName=' + user.userName)
                .then(response => response.json())
                .then(movies => {
                    if (movies.message) {
                        console.log("User does not exist");
                    } else {
                        returnUser.movies = movies;
                        dispatch(login(returnUser))
                    }
                })
        } catch (e) {
        }
    }

    // @ts-ignore
    function handleNameChange(value: string) {
        setUsername(value);
    }

    // @ts-ignore
    function handlePasswordChange(value: string) {
        setPassword(value)
    }

    return (
        <Modal size={"mini"}
            trigger={<Button style={{zIndex: '1000000'}} >Log in/Sign up</Button>}
            closeIcon>
            <Modal.Header>Log in / Sign in</Modal.Header>
            <Modal.Content>
                <Input id={"UsernameID"} autoFocus label={"Username"} onChange={(e, {value}) => handleNameChange(value)} name={"userName"} placeholder='Username'/>
            </Modal.Content>
            <Modal.Content>
                <Input id={"PasswordID"} label={"Password"} onChange={(e, {value}) => handlePasswordChange(value)} name={"password"} placeholder='Password'/>
            </Modal.Content>
            <Modal.Actions>
                <Button id={"loginButtonID"} onClick={() => onLogin({
                    userName: userName,
                    password: password,
                    movies: []
                })} type='submit'>Log in</Button>
                <Button id={"submitButtonID"} onClick={() => addUser({userName: userName, password: password, movies: []})} type='submit'>Sign up</Button>
            </Modal.Actions>
        </Modal>
    )
}

export default SignLogIn;