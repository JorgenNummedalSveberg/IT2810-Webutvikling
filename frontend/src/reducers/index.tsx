import moviesReducer from "./movies";
import genresReducer from "./genres";
import {combineReducers} from "redux";
import filterReducer from "./filters";


// Lager en samlet reducer slik at alt kan lagres i en enkelt store
const reducers = combineReducers({
    movies: moviesReducer,
    genres: genresReducer,
    filter: filterReducer
})

export default reducers;