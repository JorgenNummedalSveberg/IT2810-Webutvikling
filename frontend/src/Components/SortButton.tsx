import React,{useState} from 'react';
import './CSS/SortButton.css';

function SortButton(props:any) {
    const [isActive, toggleActive] = useState(false);
    
    function toggleSort() {
      toggleActive((isActive)  => !isActive);
      console.log("clicked");
    }

    /* Skal brukes til å laste inn en ny liste basert på sortering og filter*/
    function loadList(){

    }
    
    return (
      <div className={"SortButton"} onClick={toggleSort}>
        <svg style={{"transform":"rotate(-"+(isActive ? 90 : 0)+"deg)"}} width="258" height="452" viewBox="0 0 258 452" fill="none" xmlns="http://www.w3.org/2000/svg">
          <path d="M9.40504 203.555L203.692 9.27392C216.051 -3.09108 236.089 -3.09108 248.442 9.27392C260.796 21.6279 260.796 41.6649 248.442 54.0179L76.5281 225.927L248.437 397.83C260.791 410.189 260.791 430.224 248.437 442.578C236.083 454.937 216.046 454.937 203.687 442.578L9.40003 248.294C3.22303 242.114 0.13802 234.023 0.13802 225.928C0.138021 217.829 3.22904 209.732 9.40504 203.555Z" fill={(isActive ? "#DEDA77" : "#D1D1D1")}/>
        </svg>


        <h3 style={{"color":(isActive ? "#DEDA77" : "#D1D1D1")}}>{props.sort}</h3>
      </div>
    );
  }
  
  export default SortButton;
