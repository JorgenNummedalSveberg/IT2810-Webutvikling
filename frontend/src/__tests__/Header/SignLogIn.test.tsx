import React from 'react';
import {fireEvent, render} from '@testing-library/react';
import {Provider} from 'react-redux';
import {store} from '../../reducers/store';
import SortButton from "../../Components/Header/SortButton";
import SignLogIn from "../../Components/Header/SignLogIn";
import {DimCard} from "../../Components/MovieSection/MovieCard";

describe("Test to see if the sort button renders with our mock function, and passes props correctly", () => {
    test("Rendering of button", () => {
        const {container} = render(<Provider store={store}><SignLogIn isLogged={false}/></Provider>);
        expect(container).toMatchSnapshot();
    })
})