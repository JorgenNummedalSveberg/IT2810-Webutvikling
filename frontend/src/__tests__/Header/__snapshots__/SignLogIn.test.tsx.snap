// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test to see if the sort button renders with our mock function, and passes props correctly Rendering of button 1`] = `
<div>
  <div>
    <div
      class="makeStyles-none-2"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-loginButton-1"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Log in/Sign up
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="makeStyles-initial-3"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-loginButton-1"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Log out
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
